platforms/software/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublication.java

```diff
@@ -378,46 +368,40 @@ public MavenNormalizedPublication asNormalisedPublication() {
         populateFromComponent();
 
         // Preserve identity of artifacts
-        Map<MavenArtifact, MavenArtifact> normalizedArtifacts = normalizedMavenArtifacts();
+        Map<MavenArtifact, NormalizedMavenArtifact> normalizedArtifacts = normalizedMavenArtifacts();
 
         return new MavenNormalizedPublication(
             name,
             pom.getCoordinates(),
-            pom.getPackaging(),
+            pom.getPackaging().get(),
             normalizedArtifactFor(getPomArtifact(), normalizedArtifacts),
             normalizedArtifactFor(determineMainArtifact(), normalizedArtifacts),
             new LinkedHashSet<>(normalizedArtifacts.values())
         );
     }
 
     @Nullable
-    private static MavenArtifact normalizedArtifactFor(@Nullable MavenArtifact artifact, Map<MavenArtifact, MavenArtifact> normalizedArtifacts) {
+    private static NormalizedMavenArtifact normalizedArtifactFor(@Nullable MavenArtifact artifact, Map<MavenArtifact, NormalizedMavenArtifact> normalizedArtifacts) {
         if (artifact == null) {
             return null;
         }
-        MavenArtifact normalized = normalizedArtifacts.get(artifact);
+        NormalizedMavenArtifact normalized = normalizedArtifacts.get(artifact);
         if (normalized != null) {
             return normalized;
         }
-        return normalizedArtifactFor(artifact);
+        return new NormalizedMavenArtifact(artifact);
     }
 
-    private Map<MavenArtifact, MavenArtifact> normalizedMavenArtifacts() {
+    private Map<MavenArtifact, NormalizedMavenArtifact> normalizedMavenArtifacts() {
         return artifactsToBePublished()
             .stream()
+            .filter(Objects::nonNull)
```

Comment:
was `artifactsToBePublished` returning a set with null elements? I think we should fix that instead of filtering them out here